test/functional/byo_promises.test.js

```diff
@@ -1,36 +1,86 @@
 'use strict';
+
+const maybePromise = require('./../../lib/utils').maybePromise;
 var expect = require('chai').expect;
 
-describe('BYO Promises', function() {
-  it('should Correctly Use Blurbird promises library', {
-    metadata: {
-      requires: {
-        topology: ['single', 'ssl', 'wiredtiger']
-      }
-    },
+class CustomPromise extends Promise {}
+CustomPromise.prototype.isCustomMongo = true;
+
+const parent = { s: { promiseLibrary: CustomPromise } };
+
+describe('Optional PromiseLibrary / maybePromise', function() {
+  it('should correctly implement custom dependency-less promise', function(done) {
+    const getCustomPromise = v => new CustomPromise(resolve => resolve(v));
+    const getNativePromise = v => new Promise(resolve => resolve(v));
+    expect(getNativePromise()).to.not.have.property('isCustomMongo');
+    expect(getCustomPromise()).to.have.property('isCustomMongo');
+    expect(getNativePromise()).to.have.property('then');
+    expect(getCustomPromise()).to.have.property('then');
+    done();
+  });
+
+  it('should return a promise with extra property CustomMongo', function() {
+    const prom = maybePromise(parent, undefined, () => 'example');
+    expect(prom).to.have.property('isCustomMongo');
+    expect(prom).to.have.property('then');
+  });
+
+  it('should return a native promise with no parent', function(done) {
+    const prom = maybePromise(undefined, undefined, () => 'example');
+    expect(prom).to.not.have.property('isCustomMongo');
+    expect(prom).to.have.property('then');
+    done();
+  });
+
+  it('should return a native promise with empty parent', function(done) {
+    const prom = maybePromise({}, undefined, () => 'example');
+    expect(prom).to.not.have.property('isCustomMongo');
+    expect(prom).to.have.property('then');
+    done();
+  });
+
+  it('should return a native promise with emtpy "s"', function(done) {
+    const prom = maybePromise({ s: {} }, undefined, () => 'example');
+    expect(prom).to.not.have.property('isCustomMongo');
+    expect(prom).to.have.property('then');
+    done();
+  });
 
-    // The actual test we wish to run
+  it('should have cursor return native promise', {
+    metadata: { requires: { topology: ['single', 'ssl', 'wiredtiger'] } },
```

Comment:
What's the reason for these required metadata here? I would think this test should apply to everything.

Reply:
If I'm opening a connection in a test don't I need to specify a topology?

Reply:
If you aren't targettng a specific topology you can use the normal mocha style to target all topologies:
```
it('should have cursor return native promise', function(done) {
  // your test
});
```